<font color='Sucees'> 
### <u>REVIEWER INTRO</u>
<br />
Hi, Gaini! 🖐 My name is Andrey Kulikov and I am glad to be your reviewer in this project.<br />

You can find my comments under the heading **"<u>REVIEWER COMMENT</u>"** marked with color depending on the completeness:<br/>
<div class="alert alert-block alert-success">
<b>Success:</b> if everything is done succesfully.
</div>
<div class="alert alert-block alert-warning">
<b>Remarks:</b> if there is an area for improvement an I could give some recommendations. Yellow comment is not a hurdle to accept the work.
</div>
<div class="alert alert-block alert-danger">
<b>Needs fixing:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>

Please pay attention to not move or change my comments. It would be good to highlight your answers in some way. 
<div class="alert alert-info"> <b>Student comment:</b> For example like this.</div>
    
    
At the end of the notebook I put a colored reviewer summary block under the heading **"<u>REVIEWER SUMMARY</u>"** where give overall conclusion about your project.
    
OK, now let's go.
</font>

## Final Project - SQL

#### Goal: Use database tables to generate a value proposition for a new product.

#### Plan
1. Study the tables.
2. Make SQL queries:
   1. Task1: Find the number of books released after January 1, 2000.
   2. Task2: Find the number of user reviews and the average rating for each book.
   3. Task3: Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
   4. Task4: Identify the author with the highest average book rating (look only at books with at least 50 ratings).
   5. Task5: Identify the author with the highest average book rating (look only at books with at least 50 ratings).
3. Make conclusions

In [1]:
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
query = '''SELECT *
           FROM books 
           '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
query = '''SELECT *
           FROM authors 
           '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
query = '''SELECT *
           FROM publishers 
           '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
query = '''SELECT *
           FROM ratings 
           '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
query = '''SELECT *
           FROM reviews
           '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<font color='grenn'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-success">
Well observed data is half a job 👍👍👍
</div>
</font>

## SQL tasks

### Task 1

In [7]:
book_number = ''' SELECT 
                   count(book_id) as book_number
                   FROM books
                   WHERE publication_date >= '2000-01-01'
'''   
book_number = pd.io.sql.read_sql(book_number, con = engine)
book_number

book_number
0          821

The total books number which are released after 1st of January 2000 is 821. 

<font color='grenn'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-success">
Well done 👍👍👍
</div>
</font>

### Task 2

In [8]:
review_rating = '''SELECT 
                    reviews.book_id AS book_id,
                    books.title AS title,
                    count(reviews.review_id) AS review_number,
                    --ratings.book_id AS rating_books_id,
                    AVG(ratings.rating) AS average_rating
                FROM
                    reviews
                INNER JOIN ratings ON reviews.book_id = ratings.book_id
                INNER JOIN books ON books.book_id = ratings.book_id
                --LIMIT 3;
                GROUP BY
                    reviews.book_id,
                    books.title
                ORDER BY
                    average_rating DESC,
                    review_number DESC
'''
review_rating = pd.io.sql.read_sql(review_rating, con = engine)
review_rating

book_id                                              title  \
0         17                      A Dirty Job (Grim Reaper  #1)   
1        553            School's Out—Forever (Maximum Ride  #2)   
2        444       Moneyball: The Art of Winning an Unfair Game   
3        347  In the Hand of the Goddess (Song of the Liones...   
4        441                  Misty of Chincoteague (Misty  #1)   
..       ...                                                ...   
989      915  The World Is Flat: A Brief History of the Twen...   
990      202                                      Drowning Ruth   
991      316                  His Excellency: George Washington   
992      371                                              Junky   
993      303                               Harvesting the Heart   

     review_number  average_rating  
0               16            5.00  
1               12            5.00  
2                9            5.00  
3                6            5.00  
4                4            5.00  
..             ...             ...  
989             12            2.25  
990              9            2.00  
991              4            2.00  
992              4            2.00  
993              4            1.50  

[994 rows x 4 columns]

The table above shows the total number of reviews as well as the average rating each of the books received. A book named "A Dirty Job (Grim Reaper #1)" has the highest number of reviews and rating. The least popular book is "Harvesting the Heart". 

<font color='grenn'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-success">
Agree.
</div>
</font>

### Task 3

In [9]:
top_publishers = '''SELECT
                     books.publisher_id as publisher_id,
                     publishers.publisher,
                     count(books.book_id) as num_books
                    FROM
                     books
                    INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
                    WHERE
                     books.num_pages > 50
                    GROUP BY
                     books.publisher_id,
                     publishers.publisher
                    ORDER BY
                     num_books DESC
                    LIMIT 3; 
'''
top_publishers = pd.io.sql.read_sql(top_publishers, con = engine)
top_publishers

publisher_id                 publisher  num_books
0           212             Penguin Books         42
1           309                   Vintage         31
2           116  Grand Central Publishing         25

The table above shows top-3 publishers who published the greatest number of books which have more than 50 pages. 

<font color='grenn'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-success">
Agree one more time.
</div>
</font>

### Task 4

In [14]:
top_authors = '''SELECT
                  books.author_id AS author_id,
                  authors.author AS author,
                  rating_number.rating_num AS rating_num,
                  rating_number.average_rating AS average_rating
                  FROM 
                   (SELECT
                     book_id,
                     count(rating_id) as rating_num,
                     AVG(rating) as average_rating
                    FROM
                     ratings
                    GROUP BY
                     book_id
                   ) AS rating_number
                  INNER JOIN books ON rating_number.book_id = books.book_id 
                  INNER JOIN authors ON authors.author_id = books.author_id
                  WHERE 
                   rating_num > 50
                  ORDER BY
                   average_rating DESC
                  LIMIT 3; 

'''
top_authors = pd.io.sql.read_sql(top_authors, con = engine)
top_authors

author_id                      author  rating_num  average_rating
0        236  J.K. Rowling/Mary GrandPré          82        4.414634
1        240              J.R.R. Tolkien          74        4.391892
2        236  J.K. Rowling/Mary GrandPré          80        4.287500

<font color='orange'> 
### <u>V1 REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-warning">
The block is not fixed because I see 2 mrs Rowling in the list however should see only 1. You can get it simplier 👇
</div>
</font>

In [11]:
top_authors = '''SELECT
                  authors.author AS author,
                    COUNT(rating_id) AS rating_cnt,
                    AVG(rating) as rating_avg
                  FROM 
                    ratings
                  INNER JOIN books ON ratings.book_id = books.book_id 
                  INNER JOIN authors ON authors.author_id = books.author_id
                    GROUP BY
                    authors.author
                    HAVING 
                    COUNT(rating_id) > 50     
                  ORDER BY
                   rating_avg DESC
                  LIMIT 3; 

'''
top_authors = pd.io.sql.read_sql(top_authors, con = engine)
top_authors

author  rating_cnt  rating_avg
0         J.K. Rowling/Mary GrandPré         312    4.288462
1                    Agatha Christie          53    4.283019
2  Markus Zusak/Cao Xuân Việt Khương          53    4.264151

The table above shows the authors with the highest number of users who rated the book and the highest average rating. As can be seen J.K. Rowling/Mary GrandPré and J.R.R. Tolkien are in the top. 

<font color='red'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-danger">
Please don't group by book but only by author. We need to find the highest average rating among all books by an author.
</div>
</font>

### Task 5

In [12]:
average_text_rating = '''SELECT
                          ratings.username as username,
                          COUNT(ratings.rating_num) as average_num_text
                         FROM
                            (SELECT
                              username,
                              count(rating_id) as rating_num
                            FROM
                              ratings
                            GROUP BY
                             username
                            ) AS ratings
                         INNER JOIN reviews ON ratings.username = reviews.username
                         WHERE 
                          ratings.rating_num > 50
                         GROUP BY
                          ratings.username
'''
average_text_rating = pd.io.sql.read_sql(average_text_rating, con = engine)
average_text_rating

username  average_num_text
0          paul88                22
1          xdavis                18
2      martinadam                27
3       richard89                26
4     sfitzgerald                28
5  jennifermiller                25

The table above shows the number of text reviews each of the users who rated more than 50 books has. 

In [13]:
average_text_rating = '''with reviews as 
                         (SELECT
                           username,
                           COUNT(review_id) as review_num
                         FROM
                           reviews
                         GROUP BY
                           username
                         ),
                         ratings as
                         (SELECT
                           username,
                           COUNT(rating_id) as rating_num
                          FROM
                           ratings 
                          GROUP BY
                           username
                         )
                          SELECT
                           AVG(reviews.review_num) as average_text_review_number
                          FROM
                           reviews
                          INNER JOIN ratings ON ratings.username = reviews.username
                          WHERE
                           ratings.rating_num >= 50

'''
average_text_rating = pd.io.sql.read_sql(average_text_rating, con = engine)
average_text_rating

average_text_review_number
0                   24.222222

The average number of text reviews of users who rated more than 50 books.

<font color='grenn'> 
### <u>REVIEWER COMMENT</u>
<br/>
<div class="alert alert-block alert-success">
Perfect 👏👏👏
</div>
</font>

### Conclusion

The total books number which are released after 1st of January 2000 is 821. A book named "A Dirty Job (Grim Reaper #1)" has the highest number of reviews and rating. The least popular book is "Harvesting the Heart". Penguin Books, Vintage and Grand Central Publishinig are the most popular publishers, whereas J.K. Rowling/Mary GrandPré and J.R.R. Tolkien are in the top rated writers. The average text review's number among those users who rated more than 50 books is 24 texts.